File: cloudutil/aws/ssm.py
"""AWS SSM (Systems Manager) Parameter Store utilities."""

import os
from typing import List, Optional

from pydantic import BaseModel

from ..helper.fzf_view import FzfView
from ..utils import console
from .common import get_aws_client


class SSMParameter(BaseModel):
    """SSM Parameter model."""

    name: str
    value: str


def get_ssm_client(
    profile_name: Optional[str] = None, region_name: Optional[str] = None
):
    """Get SSM client with optional profile and region."""
    return get_aws_client("ssm", profile_name, region_name)


def list_parameters(
    prefix: str = "/",
    profile_name: Optional[str] = None,
    region_name: Optional[str] = None,
) -> List[str]:
    """List SSM parameters by path prefix."""
    ssm_client = get_ssm_client(profile_name, region_name)
    parameters = []
    next_token = None

    while True:
        kwargs: dict = {
            "Path": prefix,
            "Recursive": True,
            "WithDecryption": True,
            "MaxResults": 10,
        }
        if next_token:
            kwargs["NextToken"] = next_token

        response = ssm_client.get_parameters_by_path(**kwargs)
        for param in response["Parameters"]:
            parameters.append(param["Name"])
        next_token = response.get("NextToken")
        if not next_token:
            break

    return parameters


def get_parameter(
    name: str,
    profile_name: Optional[str] = None,
    region_name: Optional[str] = None,
) -> SSMParameter:
    """Get a specific SSM parameter by name."""
    ssm_client = get_ssm_client(profile_name, region_name)
    response = ssm_client.get_parameter(Name=name, WithDecryption=True)
    param = response["Parameter"]
    return SSMParameter(name=param["Name"], value=param["Value"])


# ── FzfView subclass ──────────────────────────────────────────────────────────


class SSMParametersView(FzfView[str]):
    """
    Interactive fzf viewer for AWS SSM Parameter Store.

    Lists parameter names, lets the user pick via fzf, then fetches and
    returns the selected ``SSMParameter`` objects.
    """

    item_type_name = "SSM parameter"
    multi_select = True

    def __init__(
        self,
        prefix: str = "/",
        profile_name: Optional[str] = None,
        region_name: Optional[str] = None,
    ) -> None:
        self._prefix = prefix
        self._profile_name = profile_name
        self._region_name = region_name
        self._selected_params: List[SSMParameter] = []

    def list_items(self) -> List[str]:
        console.print(
            f"[*] Listing SSM parameters with prefix: [bold cyan]{self._prefix}[/bold cyan]"
        )
        return list_parameters(self._prefix, self._profile_name, self._region_name)

    def item_label(self, item: str) -> str:
        return item

    def display_item(self, item: str) -> dict[str, str]:
        param = get_parameter(item, self._profile_name, self._region_name)
        self._selected_params.append(param)
        return {param.name: param.value}


# ── EC2 / SSM Session helpers (unchanged) ────────────────────────────────────


def ssm_instance(
    instance_id: str,
    tunnel: bool = False,
    document_name: str = "AWS-StartPortForwardingSessionToRemoteHost",
    remote_host: str = "",
    remote_port: int = 0,
    local_port: int = 0,
) -> None:
    if tunnel:
        if not remote_host or remote_port == 0 or local_port == 0:
            console.print(
                "[bold red][!] ERROR: Remote host, remote port, and local port "
                "are required for tunneling.[/bold red]"
            )
            return

        os.system(
            f"aws ssm start-session --target {instance_id} "
            f"--document-name {document_name} "
            f"--parameters '{{'host': ['{remote_host}'], "
            f"'portNumber': ['{remote_port}'], "
            f"'localPortNumber': ['{local_port}']}}'",
        )
    else:
        os.system(f"aws ssm start-session --target {instance_id}")


def list_ssm_instances() -> List[dict[str, str]]:
    ec2 = get_aws_client("ec2")
    response = ec2.describe_instances(Filters=[{"Name": "tag:Name", "Values": ["*"]}])
    instances = []
    for reservation in response["Reservations"]:
        for instance in reservation["Instances"]:
            if "IamInstanceProfile" not in instance:
                continue
            name = ""
            for tag in instance.get("Tags", []):
                if tag["Key"] == "Name":
                    name = tag["Value"]
                    break
            instances.append({"instance_id": instance["InstanceId"], "name": name})
    return instances


# ── FzfView subclass for EC2 instance selection ───────────────────────────────


class EC2InstanceView(FzfView[dict]):
    """
    Interactive fzf viewer for EC2 instances reachable via SSM.

    Lists instances, lets the user pick one via fzf, then starts an SSM
    session (or tunnel) to the selected instance.
    """

    item_type_name = "EC2 instance"
    multi_select = False

    def __init__(
        self,
        tunnel: bool = False,
        remote_host: str = "",
        remote_port: int = 0,
        local_port: int = 0,
    ) -> None:
        self._tunnel = tunnel
        self._remote_host = remote_host
        self._remote_port = remote_port
        self._local_port = local_port

    def list_items(self) -> List[dict]:
        return list_ssm_instances()

    def item_label(self, item: dict) -> str:
        return f"{item['instance_id']} | {item['name']}"

    def display_item(self, item: dict) -> None:
        ssm_instance(
            instance_id=item["instance_id"],
            tunnel=self._tunnel,
            remote_host=self._remote_host,
            remote_port=self._remote_port,
            local_port=self._local_port,
        )


# ── Convenience function (backwards-compatible) ───────────────────────────────


def search_parameters_with_fzf(
    prefix: str = "/",
    profile_name: Optional[str] = None,
    region_name: Optional[str] = None,
) -> None:
    """
    Search SSM parameters using fzf for interactive selection.

    Returns:
        List of selected SSMParameter objects.
    """
    SSMParametersView(
        prefix=prefix,
        profile_name=profile_name,
        region_name=region_name,
    ).run()

File: internal/awsx/ssm.go
package awsx

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"syscall"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/ec2"
	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
	"github.com/aws/aws-sdk-go-v2/service/ssm"
)

// portForwardDocument is the SSM document used for tunneling to a remote host.
const portForwardDocument = "AWS-StartPortForwardingSessionToRemoteHost"

// Parameter is one SSM Parameter Store entry.
type Parameter struct {
	Name  string
	Value string
}

// ListParameters returns every parameter name under prefix.
func ListParameters(ctx context.Context, cfg aws.Config, prefix string) ([]string, error) {
	client := ssm.NewFromConfig(cfg)
	// No WithDecryption: only names are used here, and decrypting every
	// SecureString on the path would need kms:Decrypt just to list.
	paginator := ssm.NewGetParametersByPathPaginator(client, &ssm.GetParametersByPathInput{
		Path:      aws.String(prefix),
		Recursive: aws.Bool(true),
	})

	var names []string
	for paginator.HasMorePages() {
		page, err := paginator.NextPage(ctx)
		if err != nil {
			return nil, err
		}
		for _, p := range page.Parameters {
			names = append(names, aws.ToString(p.Name))
		}
	}
	return names, nil
}

// GetParameter fetches one decrypted parameter.
func GetParameter(ctx context.Context, cfg aws.Config, name string) (Parameter, error) {
	client := ssm.NewFromConfig(cfg)
	out, err := client.GetParameter(ctx, &ssm.GetParameterInput{
		Name:           aws.String(name),
		WithDecryption: aws.Bool(true),
	})
	if err != nil {
		return Parameter{}, err
	}
	return Parameter{
		Name:  aws.ToString(out.Parameter.Name),
		Value: aws.ToString(out.Parameter.Value),
	}, nil
}

// Instance is a running EC2 instance reachable through SSM.
type Instance struct {
	ID   string
	Name string
}

// Label is the fzf line for this instance.
func (i Instance) Label() string { return i.ID + " | " + i.Name }

// ListInstances returns running, Name-tagged instances that carry an IAM
// instance profile, which is what SSM needs to reach them.
func ListInstances(ctx context.Context, cfg aws.Config) ([]Instance, error) {
	client := ec2.NewFromConfig(cfg)
	paginator := ec2.NewDescribeInstancesPaginator(client, &ec2.DescribeInstancesInput{
		Filters: []ec2types.Filter{
			{Name: aws.String("tag:Name"), Values: []string{"*"}},
			{Name: aws.String("instance-state-name"), Values: []string{"running"}},
		},
	})

	var instances []Instance
	for paginator.HasMorePages() {
		page, err := paginator.NextPage(ctx)
		if err != nil {
			return nil, err
		}
		for _, reservation := range page.Reservations {
			for _, instance := range reservation.Instances {
				if instance.IamInstanceProfile == nil {
					continue
				}
				name := ""
				for _, tag := range instance.Tags {
					if aws.ToString(tag.Key) == "Name" {
						name = aws.ToString(tag.Value)
						break
					}
				}
				instances = append(instances, Instance{
					ID:   aws.ToString(instance.InstanceId),
					Name: name,
				})
			}
		}
	}
	return instances, nil
}

// Tunnel describes an optional port-forwarding session.
type Tunnel struct {
	Enabled    bool
	RemoteHost string
	RemotePort int
	LocalPort  int
}

// StartSession opens an SSM session via the SDK and replaces this process
// with session-manager-plugin, which owns the terminal for the session's
// duration and speaks the actual data-channel protocol.
//
// session-manager-plugin is a separate install (there's no Go client for its
// protocol), but the aws CLI itself is not needed: the SDK starts the session
// and the plugin is invoked with the same argv the CLI would pass it.
func StartSession(ctx context.Context, cfg aws.Config, instanceID string, t Tunnel) error {
	binary, err := exec.LookPath("session-manager-plugin")
	if err != nil {
		return errors.New("session-manager-plugin not found in PATH — required for SSM sessions")
	}

	input := &ssm.StartSessionInput{Target: aws.String(instanceID)}
	if t.Enabled {
		if t.RemoteHost == "" || t.RemotePort == 0 || t.LocalPort == 0 {
			return errors.New("--remote-host, --remote-port and --local-port are all required for tunneling")
		}
		input.DocumentName = aws.String(portForwardDocument)
		input.Parameters = map[string][]string{
			"host":            {t.RemoteHost},
			"portNumber":      {strconv.Itoa(t.RemotePort)},
			"localPortNumber": {strconv.Itoa(t.LocalPort)},
		}
	}

	client := ssm.NewFromConfig(cfg)
	out, err := client.StartSession(ctx, input)
	if err != nil {
		return fmt.Errorf("ssm start-session failed: %w", err)
	}

	response, err := json.Marshal(sessionResponse{
		SessionId:  aws.ToString(out.SessionId),
		TokenValue: aws.ToString(out.TokenValue),
		StreamUrl:  aws.ToString(out.StreamUrl),
	})
	if err != nil {
		return fmt.Errorf("could not encode session response: %w", err)
	}
	request, err := json.Marshal(input)
	if err != nil {
		return fmt.Errorf("could not encode session request: %w", err)
	}

	// ponytail: assumes the standard AWS partition endpoint format; wrong for
	// China/GovCloud, fix by resolving via the ssm client's endpoint resolver
	// if that's ever needed.
	endpoint := fmt.Sprintf("https://ssm.%s.amazonaws.com", cfg.Region)
	argv := []string{binary, string(response), cfg.Region, "StartSession", "", string(request), endpoint}
	return syscall.Exec(binary, argv, os.Environ())
}

// sessionResponse mirrors the JSON session-manager-plugin expects as its
// first argument, which does not match ssm.StartSessionOutput's field tags.
type sessionResponse struct {
	SessionId  string `json:"SessionId"`
	TokenValue string `json:"TokenValue"`
	StreamUrl  string `json:"StreamUrl"`
}
